daemon/api_quota.go

```diff
@@ -0,0 +1,197 @@
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package daemon
+
+import (
+	"encoding/json"
+	"net/http"
+	"sort"
+
+	"github.com/snapcore/snapd/gadget/quantity"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/servicestate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/snap/naming"
+	"github.com/snapcore/snapd/snap/quota"
+)
+
+var (
+	quotaGroupsCmd = &Command{
+		Path: "/v2/quotas",
+		GET:  getQuotaGroups,
+		POST: postQuotaGroup,
+	}
+	quotaGroupInfoCmd = &Command{
+		Path: "/v2/quotas/{group}",
+		GET:  getQuotaGroupInfo,
+	}
+)
+
+type postQuotaGroupData struct {
+	// Action can be "ensure" or "remove"
+	Action    string   `json:"action"`
+	GroupName string   `json:"group-name"`
+	MaxMemory uint64   `json:"max-memory,omitempty"`
+	Parent    string   `json:"parent,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+}
+
+type quotaGroupResultJSON struct {
+	GroupName string   `json:"group-name"`
+	MaxMemory uint64   `json:"max-memory"`
+	Parent    string   `json:"parent,omitempty"`
+	Snaps     []string `json:"snaps,omitempty"`
+	SubGroups []string `json:"subgroups,omitempty"`
+}
+
+// getQuotaGroups returns all quota groups sorted by name.
+func getQuotaGroups(c *Command, r *http.Request, _ *auth.UserState) Response {
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	quotas, err := servicestate.AllQuotas(st)
+	if err != nil {
+		return InternalError("cannot get quotas: %v", err)
+	}
+
+	i := 0
+	names := make([]string, len(quotas))
+	for name := range quotas {
+		names[i] = name
+		i++
+	}
+	sort.Strings(names)
+
+	results := make([]quotaGroupResultJSON, len(quotas))
+	for i, name := range names {
+		qt := quotas[name]
+		results[i] = quotaGroupResultJSON{
+			GroupName: qt.Name,
+			Parent:    qt.ParentGroup,
+			SubGroups: qt.SubGroups,
+			Snaps:     qt.Snaps,
+			MaxMemory: uint64(qt.MemoryLimit),
+		}
+	}
+	return SyncResponse(results, nil)
+}
+
+// getQuotaGroupInfo returns details of a single quota Group.
+func getQuotaGroupInfo(c *Command, r *http.Request, _ *auth.UserState) Response {
+	vars := muxVars(r)
+	groupName := vars["group"]
+	if err := naming.ValidateQuotaGroup(groupName); err != nil {
+		return BadRequest(err.Error())
+	}
+
+	st := c.d.overlord.State()
+	st.Lock()
+	defer st.Unlock()
+
+	group, err := servicestate.GetQuota(st, groupName)
+	if err != nil {
+		return InternalError("cannot get quota: %v", err)
+	}
+	if group == nil {
+		return NotFound("cannot find quota group %q", groupName)
+	}
```

Comment:
tbh, even though I was the author of GetQuota I found myself getting confused why GetQuota was returning `err == nil` and `group == nil` leading to panics, so I'm not sure this is a great API, in the end since I needed the full set of quota groups for all the other backends, this instance right here (and some tests) is the only user of GetQuota, should we just get rid of it and always use AllQuotas()?

Reply:
Yes, sounds fine to me. I'll switch to AllQuotas and you can remove it from your PR. Thanks

Reply:
Discussed this with Samuele below, we will keep GetQuota but I've tweaked its error behavior.